### ISSUE TITLE
State Management

### ISSUE THREAD

**Issue by manssorr** [vangware/react-pair#1]
What about using state management systems like Redux, and Context API?

**Comment by loucyx** [comment]
TL;DR: You **can** use `react-pair` with both Redux and ContextAPI by creating custom hooks that access the global/shared state and then pairing them with stateless components, but in the case of Redux I don't recommend to do it.

---

The idea with [the paired hook pattern](https://知る.dev/blog/vangware/the-paired-hook-pattern-4mo0) is to help with the development of stateless components that might require a stateful behavior constantly/repeatedly, avoiding putting the state directly on the components. Generally `react-pair` is not even necessary because you can just create your custom hook and pass the props returned by that hook directly to your paired stateless component (as shown [here](https://知る.dev/blog/vangware/the-paired-hook-pattern-4mo0#the-paired-hook-pattern)). `react-pair` is only a helper function to make looping over paired components easier.

That being said, state management systems like Redux have a completely different approach for the app architecture, in which you have a "single source of truth", and all components need to be synced with that centralized state instead of having states on their own.

You could use the paired hook pattern with Redux and make custom hooks that use `useSelector`, `useDispatch` and so on, and then pair those custom hooks with your components, but being different ways of managing state (Redux embracing statefulness while Paired Hooks tries to keep away from state as much as possible), it might end up producing a really bad DX for you (you'll end up with a paired hook for every single component instead of just making your components stateful with Redux).

ContextAPI is kinda similar, but is designed to avoid prop drilling of some small state that might be needed all over your app (generally stuff like the current theme), not so much to manage the entire state of your app. I wouldn't even use it for theming nowadays because CSS became quite powerful and you can just use CSS custom properties for stuff like that. But similarly to Redux, you could create a custom hook that access the Context and returns properties that can be used by stateless components, and then pair when needed. The good thing at least is that the DX for ContextAPI with Paired Hooks is way better than Redux, because you're just using something that is designed to take care of the entire state of your app, but actually just some "global/shared" values.

Hope that helps, sorry I don't have anything better than...

![That's the neat part. You don't](https://user-images.githubusercontent.com/3979433/179418702-d5244625-8d67-4ada-b012-74bf744c9efc.png)

**Comment by manssorr** [comment]
Okay! That's clear **enough**!
**Thanks for the great explanation 😄👌**
